File: solutions/01-chronal-calibration/python/main.py
#!/usr/bin/env python3

from itertools import cycle

def main():
	
	with open("input.txt") as input_file:
		frequency_deltas = [int(n) for n in input_file.read().strip().split("\n")]
	
	print("Calibration frequency: %d" % sum(frequency_deltas))
	
	frequency_sum = 0
	known_frequencies = set()
	for delta in cycle(frequency_deltas):
		frequency_sum += delta
		if frequency_sum in known_frequencies:
			print("First repeat frequency: %d" % frequency_sum)
			break
		known_frequencies.add(frequency_sum)
	

if __name__ == "__main__":
	main()

File: solutions/01-chronal-calibration/rust/src/main.rs

use std::collections::HashSet;

fn main() {
	
	let input = include_str!("input.txt");
	
	let frequency_deltas: Vec<i32> = input.lines().map(|n| n.parse::<i32>().unwrap()).collect();
	
	println!("Calibration frequency: {}", frequency_deltas.iter().sum::<i32>());
	
	let mut frequency_sum = 0;
	let mut known_frequencies: HashSet<i32> = HashSet::new();
	for delta in frequency_deltas.iter().cycle() {
		frequency_sum += delta;
		if known_frequencies.contains(&frequency_sum) {
			println!("First repeat frequency: {}", frequency_sum);
			break;
		}
		known_frequencies.insert(frequency_sum);
	}
	
}
